component/src/lib.rs

```diff
@@ -0,0 +1,127 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::anyhow;
+use async_trait::async_trait;
+use tokio::sync::mpsc::{channel, Receiver, Sender};
+use tokio::sync::oneshot::{
+    channel as oneshotChannel, Receiver as oneshotReceiver, Sender as oneshotSender,
+};
+
+pub type IrrecoverableError = anyhow::Error;
+type JoinHandle = tokio::task::JoinHandle<()>;
+
+/// A Supervisor is instantiated to supervise a task that should always be running.
+/// A running supervisor will start a component task, and ensure that it is restarted
+/// if it ever stops.
+pub struct Supervisor<M: Manageable> {
+    irrecoverable_signal: Receiver<IrrecoverableError>,
+    cancellation_signal: Option<oneshotSender<()>>,
+    join_handle: Option<JoinHandle>,
+    manageable: M,
+}
+
+/// In order to be Manageable, a user defines the following two functions:
+///
+/// 1. A start function that launches a tokio task, as input it takes:
+/// - an irrecoverable error sender, on which the component sends information to the supervisor about
+/// an irrecoverable event that has occurred
+/// - a cancellation handle, which will be listened to in the task once an irrecoverable message
+/// has been sent, used as an "ack" that the message has been received and so the function can return
+///
+/// 2. A handle_irrecoverable which takes actions on a relaunch due to an irrecoverable error
+/// that happened. It takes the error message that may contain a stack trace and other information
+/// that was sent to the Supervisor via the tx_irrecoverable passed into start.
+#[async_trait]
+pub trait Manageable {
+    // The function that spawns a tokio task
+    async fn start(
+        &self,
+        tx_irrecoverable: Sender<anyhow::Error>,
```

Comment:
We also listen for the managed task ending because that handles when a code bug panic occurs, or anything that we didn't write code to handle. And `tx_irrecoverable` is indeed that channel we provide to force a restart. When we restart, the old receiver gets dropped along with the task, and so we initialize a new one for the new task. 

With your insight above about rearranging things so it is possible to pass mut self to handle_irrecoverable, it becomes a bit less important about which path to restart happens. But the benefit of handling an error via passing a message instead of letting it panic remains in that we can pass specific information about what happened to handle_irrecoverable for logging or fine-tuned handling.